File: app/playwright/manager.py
from typing import Optional
from PySide6.QtCore import QObject, Signal
from app.models.project import Project, PlaywrightState, ServerState
from app.process.runner import ProcessRunner
from app.server.manager import ServerManager

class PlaywrightManager(QObject):
    state_changed = Signal(object)  # PlaywrightState
    log_output = Signal(str)
    tests_finished = Signal(int)

    def __init__(self, project: Project, server_manager: ServerManager, parent=None):
        super().__init__(parent)
        self._project = project
        self._server_manager = server_manager
        self._state = PlaywrightState.IDLE
        
        self._server_manager.state_changed.connect(self._on_server_state_changed)
        
        self._runner = ProcessRunner(self)
        self._runner.output_ready.connect(self._on_runner_output)
        self._runner.error_ready.connect(self._on_runner_error)
        self._runner.process_finished.connect(self._on_runner_finished)
        self._runner.process_error.connect(self._on_runner_error)
        
        self._report_runner = ProcessRunner(self)
        self._report_runner.output_ready.connect(self._on_runner_output)
        self._report_runner.error_ready.connect(self._on_runner_error)
        self._report_runner.process_error.connect(self._on_runner_error)
        
        self._pending_command: Optional[str] = None

    @property
    def state(self) -> PlaywrightState:
        return self._state

    def update_project(self, project: Project):
        self._project = project

    def run_tests(self):
        self._execute_or_wait_for_server(self._project.playwright.command)

    def run_ui(self):
        self._execute_or_wait_for_server(self._project.playwright.ui_command)

    def run_debug(self):
        self._execute_or_wait_for_server(self._project.playwright.debug_command)

    def show_report(self):
        cmd = self._project.playwright.report_command
        if not cmd:
            self.log_output.emit("No report command configured")
            return
            
        self.log_output.emit(f"Executing report command: {cmd}")
        self._report_runner.start(cmd, self._project.path)

    def stop(self):
        if self._runner.is_running():
            self._runner.stop()
        self._pending_command = None
        self._set_state(PlaywrightState.IDLE)

    def _execute_or_wait_for_server(self, command: str):
        if not command:
            self.log_output.emit("Command is empty")
            return
            
        if self._state == PlaywrightState.RUNNING:
            self.log_output.emit('Playwright is already running')
            return
            
        if not self._project.playwright.enabled:
            self.log_output.emit('Playwright is not enabled')
            return

        if self._project.server.enabled and self._server_manager.state != ServerState.RUNNING:
            self._pending_command = command
            self._set_state(PlaywrightState.STARTING)
            
            try:
                self._server_manager.ready.disconnect(self._on_server_ready)
            except RuntimeError:
                pass
                
            self._server_manager.ready.connect(self._on_server_ready)
            self._server_manager.start()
            self.log_output.emit('Waiting for server to be ready...')
        else:
            self._run_command(command)

    def _on_server_ready(self):
        try:
            self._server_manager.ready.disconnect(self._on_server_ready)
        except RuntimeError:
            pass
            
        if self._pending_command is not None:
            cmd = self._pending_command
            self._pending_command = None
            self._run_command(cmd)

    def _on_server_state_changed(self, server_state: ServerState):
        if self._pending_command is not None and server_state in (ServerState.ERROR, ServerState.STOPPED):
            try:
                self._server_manager.ready.disconnect(self._on_server_ready)
            except RuntimeError:
                pass
            self._pending_command = None
            self._set_state(PlaywrightState.ERROR)
            self.log_output.emit('Server failed to start or timed out. Playwright execution cancelled.')

    def _run_command(self, command: str):
        self._set_state(PlaywrightState.RUNNING)
        self.log_output.emit(f"Running command: {command}")
        
        extra_env = {}
        if self._project.server.enabled and self._server_manager.active_url:
            extra_env["BASE_URL"] = self._server_manager.active_url
            extra_env["PLAYWRIGHT_TEST_BASE_URL"] = self._server_manager.active_url
            if self._server_manager.active_port > 0:
                extra_env["PORT"] = str(self._server_manager.active_port)
                
        self._runner.start(command, self._project.path, extra_env=extra_env)

    def _set_state(self, new_state: PlaywrightState):
        if self._state != new_state:
            self._state = new_state
            self.state_changed.emit(new_state)

    def _on_runner_output(self, text: str):
        self.log_output.emit(text)

    def _on_runner_error(self, text: str):
        self.log_output.emit(f'[ERROR] {text}')

    def _on_runner_finished(self, exit_code: int, status: str):
        if exit_code == 0:
            self._set_state(PlaywrightState.PASSED)
        else:
            self._set_state(PlaywrightState.FAILED)
            
        self.tests_finished.emit(exit_code)
        self.log_output.emit(f"Playwright process finished with exit code {exit_code} ({status})")

File: devmanager-app/internal/playwright/manager.go
// Package playwright porta app/playwright/manager.py.
//
// Orquesta la ejecuci├│n de pruebas Playwright con auto-start del servidor:
// si el servidor est├í habilitado y no RUNNING, guarda el comando pendiente,
// arranca el servidor y espera su se├▒al ready (listener) para ejecutar.
package playwright

import (
	"strconv"
	"sync"

	"github.com/d-l-n/devmanager/internal/models"
	"github.com/d-l-n/devmanager/internal/process"
	"github.com/d-l-n/devmanager/internal/server"
)

type State string

const (
	StateIdle     State = "idle"
	StateStarting State = "starting"
	StateRunning  State = "running"
	StatePassed   State = "passed"
	StateFailed   State = "failed"
	StateError    State = "error"
)

type Callbacks struct {
	OnStateChange func(state State)
	OnLog         func(line string, isError bool)
	OnFinished    func(exitCode int) // paridad tests_finished
}

type Manager struct {
	mu sync.Mutex

	project models.Project
	server  *server.Manager
	state   State

	pendingCommand string
	stopRequested  bool

	runner       *process.Runner
	reportRunner *process.Runner
	cb           Callbacks
}

func NewManager(project models.Project, srv *server.Manager, cb Callbacks) *Manager {
	m := &Manager{
		project: project,
		server:  srv,
		state:   StateIdle,
		cb:      cb,
	}

	m.runner = process.NewRunner(process.RunnerCallbacks{
		OnStdout:   func(line string) { m.log(line, false) },
		OnStderr:   func(line string) { m.log("[ERROR] "+line, true) },
		OnFinished: m.onRunnerFinished,
		OnError:    func(desc string) { m.log("[ERROR] "+desc, true) },
	})

	// Paridad Python: el report runner NO conecta process_finished;
	// show-report vive hasta que el usuario lo cierra.
	m.reportRunner = process.NewRunner(process.RunnerCallbacks{
		OnStdout: func(line string) { m.log(line, false) },
		OnStderr: func(line string) { m.log("[ERROR] "+line, true) },
		OnError:  func(desc string) { m.log("[ERROR] "+desc, true) },
	})

	// Observaci├│n del servidor (equivalente a connect(state_changed)/connect(ready)).
	srv.AddStateListener(m.onServerStateChanged)
	srv.AddReadyListener(m.onServerReady)
	return m
}

func (m *Manager) State() State {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.state
}

func (m *Manager) UpdateProject(p models.Project) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.project = p
}

func (m *Manager) log(line string, isError bool) {
	if m.cb.OnLog != nil {
		m.cb.OnLog(line, isError)
	}
}

func (m *Manager) setState(s State) {
	m.mu.Lock()
	if m.state == s {
		m.mu.Unlock()
		return
	}
	m.state = s
	onChange := m.cb.OnStateChange
	m.mu.Unlock()
	if onChange != nil {
		onChange(s)
	}
}

func (m *Manager) RunTests() { m.executeOrWaitForServer(m.project.Playwright.Command) }
func (m *Manager) RunUI()    { m.executeOrWaitForServer(m.project.Playwright.UICommand) }
func (m *Manager) RunDebug() { m.executeOrWaitForServer(m.project.Playwright.DebugCommand) }

// ShowReport replica show_report.
func (m *Manager) ShowReport() {
	cmd := m.project.Playwright.ReportCommand
	if cmd == "" {
		m.log("No report command configured", false)
		return
	}
	m.log("Executing report command: "+cmd, false)
	if err := m.reportRunner.Start(cmd, m.project.Path, nil); err != nil {
		// Runner ya notific├│ por OnError; aqu├¡ no cambia estado (paridad).
		_ = err
	}
}

// Stop replica stop(): arma stopRequested SIEMPRE para que un finish tard├¡o
// del runner moribundo no cuente como resultado de tests.
func (m *Manager) Stop() {
	m.mu.Lock()
	m.stopRequested = true
	m.pendingCommand = ""
	m.mu.Unlock()
	if m.runner.IsRunning() {
		m.runner.Stop()
	}
	m.setState(StateIdle)
}

// executeOrWaitForServer replica _execute_or_wait_for_server.
func (m *Manager) executeOrWaitForServer(command string) {
	if command == "" {
		m.log("Command is empty", false)
		return
	}
	if m.State() == StateRunning {
		m.log("Playwright is already running", false)
		return
	}
	m.mu.Lock()
	project := m.project
	m.mu.Unlock()

	if !project.Playwright.Enabled {
		m.log("Playwright is not enabled", false)
		return
	}

	if project.Server.Enabled && m.server.State() != models.StateRunning {
		m.mu.Lock()
		m.pendingCommand = command
		m.mu.Unlock()
		m.setState(StateStarting)
		m.server.Start() // no-op si ya STARTING (guard interno)
		m.log("Waiting for server to be ready...", false)
		return
	}
	m.runCommand(command)
}

// onServerReady replica _on_server_ready.
func (m *Manager) onServerReady() {
	m.mu.Lock()
	pending := m.pendingCommand
	m.pendingCommand = ""
	m.mu.Unlock()
	if pending != "" {
		m.runCommand(pending)
	}
}

// onServerStateChanged replica _on_server_state_changed: cancela si el
// servidor cae a ERROR/STOPPED con comando pendiente.
func (m *Manager) onServerStateChanged(s models.ServerState) {
	m.mu.Lock()
	pending := m.pendingCommand
	m.mu.Unlock()
	if pending == "" {
		return
	}
	if s == models.StateError || s == models.StateStopped {
		m.mu.Lock()
		m.pendingCommand = ""
		m.mu.Unlock()
		m.setState(StateError)
		m.log("Server failed to start or timed out. Playwright execution cancelled.", false)
	}
}

// runCommand replica _run_command.
func (m *Manager) runCommand(command string) {
	m.mu.Lock()
	m.stopRequested = false
	project := m.project
	m.mu.Unlock()

	m.setState(StateRunning)
	m.log("Running command: "+command, false)

	extraEnv := map[string]string{}
	if project.Server.Enabled && m.server.ActiveURL() != "" {
		activeURL := m.server.ActiveURL()
		extraEnv["BASE_URL"] = activeURL
		extraEnv["PLAYWRIGHT_TEST_BASE_URL"] = activeURL
		if port := m.server.ActivePort(); port > 0 {
			extraEnv["PORT"] = strconv.Itoa(port)
		}
	}

	if err := m.runner.Start(command, project.Path, extraEnv); err != nil {
		// OnError ya registr├│ el fallo en el log (paridad process_error).
		_ = err
	}
}

// onRunnerFinished se basa en _on_runner_finished.
//
// DIVERGENCIA INTENCIONAL vs Python (#8): en Python, tras Stop() el runner
// moribundo (taskkill => exit no-0) marca StateFailed y emite tests_finished,
// lanzando notificaci├│n de tests fallidos aunque el usuario simplemente haya
// cancelado. Aqu├¡ mantenemos un cierre limpio: si el usuario par├│, terminamos
// en StateIdle y NO emitimos OnFinished/tests_finished. Decisi├│n expl├¡cita de
// producto para evitar la notificaci├│n de fallo cosm├®tico.
func (m *Manager) onRunnerFinished(exitCode int, status string) {
	m.mu.Lock()
	stopped := m.stopRequested
	m.mu.Unlock()

	if stopped {
		m.mu.Lock()
		m.stopRequested = false
		m.mu.Unlock()
		m.setState(StateIdle)
		m.log("Playwright run stopped by user (exit code "+strconv.Itoa(exitCode)+").", false)
		return
	}

	if exitCode == 0 {
		m.setState(StatePassed)
	} else {
		m.setState(StateFailed)
	}
	if m.cb.OnFinished != nil {
		m.cb.OnFinished(exitCode)
	}
	m.log("Playwright process finished with exit code "+strconv.Itoa(exitCode)+" ("+status+")", false)
}
